### PR TITLE
Add Express Checkout Element (ECE) support to shortcode cart and checkout

### LINKED ISSUE #8870: Integrate ECE for Payment Processing in the Shortcode Cart/Checkout
Implement the Stripe Express Checkout Element (ECE) to process payments on the WooCommerce shortcode cart and checkout pages. 

### Acceptance criteria:

- The ECE implementation must successfully process payments on the WooCommerce shortcode cart and checkout pages without disrupting the existing checkout flow.
- Ensure there are no regressions in Payment Request Buttons (PRB) when ECE is disabled.
- There are no regressions when no express checkout payment methods are active.

### PR DESCRIPTION
**Pull request by reykjalin**
Fixes #8870 

## Changes proposed in this Pull Request

- Renders ECE elements on Shortcode Cart and Checkout pages when the `_wcpay_feature_stripe_ece` option is set to `1`.

## Testing instructions

*

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8895` or branch name `add/8870-ece-for-shortcode-cart-and-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/8870-ece-for-shortcode-cart-and-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 0728307d457b8ba2b5f1552cf8e10c9fc827c650
- Build time: 2024-06-07 04:50:01 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/express-checkout/index.js [hunk 1]
```diff
@@ -185,41 +189,70 @@ jQuery( ( $ ) => {
...
+					// FIXME: This is a total hack, we need some way to get the shipping information before rendering the button.
+					//        Possible to just send an empty address maybe? Just make up a free shipping thing since the shipping
+					//        address change event is sent as soon as the payment sheet pops up?
+					shippingRates: options.displayItems
+						.filter( ( i ) => i.label === 'Shipping' )
+						.map( ( i ) => ( {
+							id: `rate-${ i.label }`,
+							amount: i.amount,
+							displayName: i.label,
+						} ) ),
```
Comment (reykjalin) [inline client/express-checkout/index.js]: I'm at a bit of a loss here, not sure how to get the currently selected address, and not even sure what's supposed to happen during a guest session when there is no address available before clicking the button.

Either way the payment sheet UI immediately triggers the `shippingaddresschange` event as soon as it loads, so this might not be a huge problem? Except maybe on slow networks where it's possible that the initial payment method shows up at a lower price than the actual options?

Another thing to consider is what happens when the response is empty. If there is no shipping method available for the currently loaded address then the GPay button will fail to load because if `shippingAddressRequired` is `true` you **must** initialize with a list of `shippingRates`; there is no way to defer loading them until later.

I'm leaning towards:

1. Use the current method to fall back on the "Shipping" display item.
    - This is a little bit brittle in the sense that it could not be present if there are no shipping methods configured for the customer's currently loaded address.
2. If there is no "Shipping" item in the display items: don't try to load the button.

This has the downside of making it so that the button won't load even if the customer is on the checkout page and changes their address to a place that has configured shipping methods. We can work around that by adding listeners to the address inputs on the checkout page but that might be a whole separate area of testing if different themes load those inputs in different ways.

What are your thoughts here @bborman22 in terms of where we should focus the effort here? Maximum compatibility or just the "happy path" of customer has an address loaded that has a corresponding shipping method?

File: client/express-checkout/index.js [hunk 2]
```diff
@@ -185,41 +189,70 @@ jQuery( ( $ ) => {
...
+				);
+
+				if ( response.result === 'success' ) {
+					elements.update( { amount: response.total.amount } );
+					event.resolve( {
+						shippingRates: response.shipping_options,
+					} );
+				} else if ( response.result === 'fail' ) {
+					event.reject();
+				}
```
Comment (reykjalin) [inline client/express-checkout/index.js]: It might be prudent to try to revert the changes here in some way, for example to reset the amount and/or selected shipping method, and something similar.

File: client/express-checkout/index.js [hunk 3]
```diff
@@ -185,41 +189,70 @@ jQuery( ( $ ) => {
...
+					elements.update( { amount: response.total.amount } );
+					event.resolve( {
+						lineItems: response.displayItems.map( ( i ) => ( {
+							...i,
+							name: i.label,
+						} ) ),
+					} );
+				} else if ( response.result === 'fail' ) {
+					event.reject();
+				}
```
Comment (reykjalin) [inline client/express-checkout/index.js]: It might be prudent to try to revert the changes here in some way, for example to reset the amount and/or selected shipping method, and something similar.

File: includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php [hunk 4]
```diff
@@ -93,6 +95,57 @@ public function ajax_get_shipping_options() {
...
+
+		$product_view_options      = filter_input_array( INPUT_POST, [ 'is_product_page' => FILTER_SANITIZE_SPECIAL_CHARS ] );
+		$should_show_itemized_view = ! isset( $product_view_options['is_product_page'] ) ? true : filter_var( $product_view_options['is_product_page'], FILTER_VALIDATE_BOOLEAN );
+
+		$data           = [];
+		$data          += $this->express_checkout_button_helper->build_display_items( $should_show_itemized_view );
+		$data['result'] = 'success';
+
+		wp_send_json( $data );
+	}
```
Comment (reykjalin) [inline includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:147]: These handlers are copied verbatim from `includes/class-wc-payments-payment-request-button-handler.php`, aside from minor tweaks to use `$this->express_checkout_button_helper` where appropriate.

I think this copy-paste is fine, but maybe we should try to attempt some sort of re-use? I'm leaning towards leaving this as is though.

File: client/express-checkout/index.js (hunk 1 above)
Comment (reykjalin) [reply client/express-checkout/index.js]: cc @rafaelzaleski just in case you faced something similar with the Blocks integration, although I think you always have an address provided by the Blocks API, so you might not have run into this 🤔

Comment (reykjalin) [reply client/express-checkout/index.js]: And I forgot to mention: I think you may get shipping rates from the backend if you send the API an empty address _so long as there is a "rest of world" shipping method configured_. I'm not sure what happens if there is no "rest of world" shipping method configured.

Comment (reykjalin) [reply client/express-checkout/index.js]: I've implemented the approach I'm referring to in 63dabfd6cfcf27b76010f77a9baa8312bf736d55.

Comment (bborman22) [reply client/express-checkout/index.js]: Just wanted to comment that I'm taking a look here to get a better grasp. Assuming the approach in 63dabfd unblocks progress here, let's move forward with that and create an issue to track any further digging needed. I was trying to take a look and didn't see a 1 to 1 mapping with PRB in this flow, but to summarize we should be targeting feature parity between ECE and PRB if that helps answer the question at all as I'm not sure what happens in those scenarios with PRB yet and will be digging in now to find out.

Comment (reykjalin) [reply client/express-checkout/index.js]: The PRB (current) implementation doesn't require the address upfront to load the buttons, you're instead allowed to wait for the `shippingaddresschange` event to get proper shipping rates, so this is a new problem for ECE.

I think this idea of not rendering the button when we can't find shipping rates is a reasonable middle-ground to "solve" the issue caused by the new API.

**Comment by reykjalin** [comment]
I tested "rest of world" here by:

1. Add a rest of world shipping method.
2. Process a payment via GPay using an address that falls into the "Rest of world" shipping zone.
3. Remove the rest of world shipping method.
4. Add a product to the cart, go back to the cart page, and click GPay button.
5. No address is selected because the shipping method is invalid.

So it seems we don't need to think about "what if there are no shipping rates" more than we already are 🙌

File: client/express-checkout/index.js [hunk 5]
```diff
@@ -177,50 +188,146 @@ jQuery( ( $ ) => {
...
+					shippingAddressRequired: options.requestShipping,
+					phoneNumberRequired: options.requestPhone,
+					shippingRates,
 				};
+				wcpayECE.block();
 				event.resolve( clickOptions );
 			} );
 
+			// FIXME: This handler is copied from ./event-handlers.js. We should re-use the same function here.
+			eceButton.on( 'shippingaddresschange', async ( event ) => {
```
Comment (asumaran) [inline client/express-checkout/index.js]: @reykjalin I think we are duplicating efforts. Check my PR https://github.com/Automattic/woocommerce-payments/pull/8914/files#diff-71dc9212813424406f6e197c711b7b1776d744686daba30456192cfdadbe99b2R224-R226

Comment (asumaran) [reply client/express-checkout/index.js]: I'm re-using the handlers defined in `event-handlers.js`. I thought it was meant to be re-used? @rafaelzaleski

Comment (reykjalin) [reply client/express-checkout/index.js]: Oh we probably are! Although I'm not sure passing the elements in works since the blocks implementation doesn't seem to have access to those 🤔

https://github.com/Automattic/woocommerce-payments/blob/95b74a487a7811604097acfd67dac428bfc0e149/client/express-checkout/blocks/components/express-checkout-component.js#L15-L50

I think the blocks implementation needs to update the `buttonOptions`? Maybe? Oh or maybe just expose the elements in the `useExpressCheckout` hook I guess 🤔 

I'll work it into my PR so I can incorporate your changes as well 👍

Comment (asumaran) [reply client/express-checkout/index.js]: Maybe we should use the same branch? Or at least share the same changes in both PRs

Comment (reykjalin) [reply client/express-checkout/index.js]: I added my take on this in 0973d1f2193281f5d3288bb319720d3728742937

**Comment by reykjalin** [comment]
Closing this PR in favor of https://github.com/Automattic/woocommerce-payments/pull/8914 due to the amount of duplicated code between the 2.

File: client/express-checkout/index.js (hunk 5 above)
Comment (reykjalin) [reply client/express-checkout/index.js]: Merged this PR into yours

File: client/express-checkout/index.js (hunk 3 above)
Comment (reykjalin) [reply client/express-checkout/index.js]: Actually, I think the payment sheet does this automatically, so I don't think there's any action required here.

File: client/express-checkout/index.js (hunk 2 above)
Comment (reykjalin) [reply client/express-checkout/index.js]: Actually, I think the payment sheet does this automatically, so I don't think there's any action required here.

**Comment by bborman22** [comment]
> No address is selected because the shipping method is invalid.

Could you expand on this a bit? Does the pay sheet display in that case? Does your conditional prevent the button from rendering? I did some digging on this today and found that we might be able to leverage what the product page is doing and provide a "Pending" shipping method of amount 0 until the shopper selects an address. I also noticed that there is some geo lookup happening to provide a rough guide to inform shipping rates so could cause some unnecessary prevention of express checkout (shopping from CA according to geo lookup, but will be shipping to NY which would have rates).

~~Unfortunately the messaging is not great (at least in Google Pay) when the address doesn't have shipping options, I wish it was a little more clear that there isn't a problem with the address and that it just doesn't ship to that location.~~
Actually this messaging is coming from us, let's update this to be more clear if possible, I'll create an issue to track some of these thoughts. No need to address any of this as part of the shortcode / product PR in #8914.
**Edit Take 2: I guess it's not our messaging, just the error code we respond with. Seems the event.reject sets this messaging.**

Screenshot 2024-06-07 at 11 43 00 AM

**Comment by reykjalin** [comment]
> Could you expand on [No address is selected because the shipping method is invalid] a bit?

Sure! The payment sheet pops up with the address list expanded, prompting you to select an address before you can click "pay". If you select an address that doesn't have an associated shipping method you'll get an "invalid shipping address" error and have to select a different address.

> we might be able to leverage what the product page is doing and provide a "Pending" shipping method of amount 0 until the shopper selects an address

Stripe's ECE API [doesn't seem to support that](https://docs.stripe.com/js/element/events/on_click?type=expressCheckoutElement#element_on_click-handler-resolve-shippingRates). I know it's supported by Google Pay themselves but it doesn't look like Stripe exposes that as an option.

> I guess it's not our messaging, just the error code we respond with. Seems the event.reject sets this messaging.

Yup, and AFAICT [there is no way to change the error message](https://docs.stripe.com/js/elements_object/express_checkout_element_shippingaddresschange_event#express_checkout_element_on_shipping_address_change-handler-reject) to give the customer a better idea of what's wrong.
